### ISSUE TITLE
Matmul throws CL_MEM_OBJECT_ALLOCATION_FAILURE on OpenCL

### ISSUE THREAD

**Issue by villekf** [arrayfire/arrayfire#2341]
The following code

```
int size1 = 400;
int size2 = 2100000;
    
array HL = constant(0.f, size1, size1, f32);
    
for (int kk = 0; kk < 2; kk++) {

    array LH = constant(0.f, size1 * size2, 1, f32);
        
    LH = moddims(LH, size1, size2);
        
    HL = matmulNT(LH, LH);
    
}
```

throws 
CLBlast: OpenCL error: clEnqueueNDRangeKernel: -4
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Internal error:998):
In function opencl::Array<T> opencl::matmul(const opencl::Array<T>&, const opencl::Array<T>&, af_mat_prop, af_mat_prop) [with T = float]
In file src/backend/opencl/blas.cpp:120
CLBlast Error (-4): CL_MEM_OBJECT_ALLOCATION_FAILURE 
on OpenCL. If I remove the loop, everything works fine. Also, the above code runs fine (and uses the correct amount of memory) on CUDA backend.

ArrayFire v3.7.0 (OpenCL, 64-bit Linux, build d25ab30)
[0] NVIDIA: Tesla P100-PCIE-16GB, 16280 MB
-1- NVIDIA: Quadro K620, 2001 MB
-2- INTEL: Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz, 257831 MB

**Comment by 9prady9** [comment]
I have much less capable card and I was able to reproduce similar behavior with 1.1 million elements along dim 0. Thank you for reporting it. We will post an update as soon as we have something to share.

**Comment by 9prady9** [comment]
I have much less capable card and I was able to reproduce similar behavior with 1.1 million elements along dim 0. Thank you for reporting it. We will post an update as soon as we have something to share.

**Comment by 9prady9** [comment]
Quick update: Found that clblas doesn't have this issue. Only when blas upstream is CLBlast, this issue surfaces.

**Comment by 9prady9** [comment]
As per [suggestions](https://github.com/CNugteren/CLBlast/issues/333#issuecomment-439207745) from @CNugteren I have experimented with the CLBlast API and came to the following conclusion/diemma.

**Suggestion (1)** - Doesn't help the use case since the queried temporary buffer still has large size when indirectGemm route is taken.

**Suggestion (2)** -  Has the same memory restriction as (1) and requires pre-transposed inputs. I am not sure having inputs pre-transposed for the API `matmulNT` is the right way to do it. @umar456 @pavanky Any thoughts ?

**Suggestion (3)** - Changing the minimum indirect buffer size does help but I am not sure how to determine this limit at runtime for a given device and it's max memory. @CNugteren Any ideas ? I couldn't understand what governs [this check](https://github.com/CNugteren/CLBlast/blob/43e3f27254c4f7e4a0b332f5b88965c53c20bdd1/src/routines/level3/xgemm.hpp#L70). Is the current limit for performance(local/global reads/writes) reasons or a tuned parameter for a given device ?

**Comment by CNugteren** [comment]
Yes the first two suggestions **(1) and (2)** are perhaps more for individual users to do rather than for you as a library developer to solve things. They can be useful or good to know still for the user, e.g. if he/she would have allocated a 1024x1024 matrix rather than a 1023x1023 matrix that would've helped for example. And pre-transposing is something that might require a temporary buffer, but that depends on what you did before of course, e.g. you could initialize/compute your values in the way that CLBlast likes them.

About **(3)**: there are basically two approaches to do GEMM, either in a single kernel which can't make any assumptions, or using pre/post processing kernels and a GEMM kernel that does make assumptions (e.g. `m` is divisible by 64). The latter is typically the faster kernel, especially for larger matrices. But this is device dependent of course, so there is a tuning parameter that decides where this cut-off lies: either run the in-direct single kernel (for matrices below a certain size) or run the multi-kernel approach with temporary matrices. See [here](https://github.com/CNugteren/CLBlast/blob/43e3f27254c4f7e4a0b332f5b88965c53c20bdd1/doc/details_gemm.md) for more details.

So if you don't want temporary buffers at all, you can change this tuning parameter and set it such that in-direct kernel is always used. You do pay in terms of performance though. By the way, `clBLAS` has the same design with a temporary buffer. The only difference is that is requires the user to allocate that buffer, whereas in CLBlast this is optional.

**Comment by 9prady9** [comment]
thank you @CNugteren 

(1) and (2) falls in line with my earlier conclusions.

For (3), from what you said and I read from the CLBlast paper, setting the cut off much higher has only one side-effect, albeit a heavy one, a potential performance drop. Yes, that is the reason our clBLAS backend doesn't experience the memory allocation exception for same problem size - since it is sacrificing the performance for successful execution - their design choice.

I unable to see a straightforward/programmatic(performance oriented) way to choose this cut off that works for most cases. I think switching to less performant direct kernel has to be user's choice instead of ArrayFire doing it automatically. Perhaps something like override parameters API of CLBLast. This parameter would do nothing for non-CLBlast backends as expected. @umar456 @pavanky Any suggestions ?

**Comment by CNugteren** [comment]
What we can also consider is that CLBlast itself would always first check if there is sufficient memory available to allocate the temporary buffers and only then use the faster kernel. But I'm not sure how big the overhead is for querying the total amount of free memory.

**Comment by 9prady9** [comment]
Yes, I did think about that and it can be done fairly quickly inside ArrayFire since we do have a memory manager which keeps track of total memory used. However, I am thinking about the following: _Automatically choosing a high performant code path is desirable_ vs _Automatically choosing a potentially less performant kernel due to memory constraint_ - I am not sure if it is a good or bad idea.

**Comment by 9prady9** [comment]
@villekf As you can see from our earlier comments, the fix for this bug/issue is not a straightforward one. A temporary fix would be to increase the threshold used by CLBlast but that would potentially cause issues in different setup. A more permanent solution is to change few things in our memory management and that will take some time. Therefore, we will push a fix for this in later releases. We will also update this issue when we have something to share. Thanks once again for reporting this!
